certification-content-bundle-spec.md

```diff
@@ -0,0 +1,112 @@
+
+tags:
+product_tags:
+role_tags:
+domain_tags:
+
+credits: 80
+
+steps: ...
+```
+
+### Certification attributes
+
+attribute          | required | type       | notes
+-------------------| -------- | ---------- | -----------------------------------------
+default_locale     | ✓        | string     | Corresponds to the locale that the certification is authored in. Authoring tools can use this as a hint to notify localizers when content in the default locale is updated. Also, it provides a hint to the learner interface about which locale to display if an instruction/resource is not localized for the learner's current locale.
+schema_version     | ✓        | integer    | Which version of the certification bundle schema you are using
+title              | ✓        | dictionary | A locale dictionary of the certification title, such as "Virtual Machines in GCP"
+description        |          | dictionary | A locale dictionary of the certification description which describes the contents of the certification
+objectives         |          | dictionary | A locale dictionary of objective arrays
+audience           |          | dictionary | A locale dictionary of audiences
+prerequisites      |          | dictionary | A locale dictionary of prerequisites
+tags               |          | array      | Array of strings to be used as hints in searching, etc.
+product_tags       |          | array      | Array of strings from the "Products" column in [this sheet](https://docs.google.com/spreadsheets/d/1hUUch85HBRsRJsgRo9VCg0Pn7ZXi21sl6JU7VOr9LP8)
+role_tags          |          | array      | Array of strings from the "Roles" column in [this sheet](https://docs.google.com/spreadsheets/d/1hUUch85HBRsRJsgRo9VCg0Pn7ZXi21sl6JU7VOr9LP8)
+domain_tags        |          | array      | Array of strings from the "Domain" column in [this sheet](https://docs.google.com/spreadsheets/d/1hUUch85HBRsRJsgRo9VCg0Pn7ZXi21sl6JU7VOr9LP8)
+credits            |          | integer    | Price of the certification
+steps              | ✓        | array      | An array of `steps` (see [Steps](#steps) below for details)
+
+### Steps
+
+A certification is composed of multiple `steps`. A step can be either a [CourseTemplate](./course-template-bundle-spec.md) or an [Exam](./exam-bundle-spec.md) that the user must complete to progress through the certification.
+
+#### Gated Steps
+
+Steps can optionally be marked as "gated". A gated step and all of its succeeding steps will be unavailable to the user until all preceding steps are completed.
+
+For example, consider the following order of steps, where steps `C` and `E` are gated:
+
+```
+A -- B -- |C| -- D -- |E|
+```
+
+- At first, only `A` and `B` are available to the user.
+- Once `A` and `B` are completed (in any order), `C` and `D` become available.
+- Once `C` and `D` are completed (in any order), `E` becomes available.
+
+#### Course Step Example
+
+```yml
+type: course_template
+id: qwiklabs-test-content/my-cool-course
```

Comment:
@jgan can you please indicate here how the "version" would look?

Reply:
Would it maybe look this?

```
id: qwiklabs-test-content/my-cool-course/1.2.3
```

Reply:
Here's what it looks like coming back from Alexandria

"libraries/gcp-it-cert-content/content/course_1/versions/1.0"

https://www.qwiklabs.com/admin/alexandria_subscriptions/libraries%2Fgcp-it-cert-content/content/libraries%2Fgcp-it-cert-content%2Fcontent%2Fcourse_1%2Fversions%2F1.0

So in general it has major.minor, but we would need to enforce that up and down the chain.

Reply:
Sure, but that rule should be defined and enforced with a different specification / PR. Application-level semantic version format enforcement is beyond the scope of this spec, which is just for defining Certifications. I'll add this expected format here though.

`libraries/gcp-it-cert-content/content/course_1/versions/1.0` is a little verbose. Is there any reason that can't be `gcp-it-cert-content/course_1/1.0`?

Reply:
Just FYI. Confirmed with @akefeli that in Alexandria the "semantic_version" has 0 or 1 periods and that the elements on either side of the period should be a non-negative integer

Reply:
After discussion with @shoffing and @akefeli to make parsing and validation much simpler we decided to separate the version with a different delimiter `@`. So the first `/` denotes the library. Subsequent `/` are part of the slug and the sematic_version is after `@`
For example
```
type: exam
id: qwiklabs-test-content/my-excellent-exam@1.0
```

Reply:
I don't think `/` is a valid character for slugs, so we can probably just use slashes for both delimiters. I'd rather not introduce a new convention here that isn't used anywhere else.

Reply:
`/` is a valid character in slugs, which makes it hard to use slashes as both delimiters. Slashes are URL-escaped (CGI) coming from Alexandria, then unescaped in website: https://github.com/CloudVLab/qwiklab-website/blob/master/app/services/alexandria_service.rb#L141-L151

Reply:
Huh - that really surprises me. I thought that the whole reason we were using slashes as delimiters in content IDs is because we don't allow slashes in libraries, slugs, etc. Did that change at some point? FWIW it doesn't look like any existing labs have a slash in the slug.

Reply:
This code has been in website since about a year ago, and I don't know about the state of things before it. Here's the PR: https://github.com/CloudVLab/qwiklab-website/pull/3272

Reply:
@akefeli @davetchen @jgan do you know why we try to allow slashes in slugs? I was under the impression that we did not allow that. I think we should disallow that and use slashes as delimiters.

Reply:
I just got done chatting with Dave at our standup. So the slashes in slugs are currently being used by git whisperer to categorize the content in directories in git. Authors had a hard time when there were hundreds of labs under a single flat directory. Some authors would use namespacing with dashes or underscores or something (like `steves-cool-labs_goto-is-awesome`) but it wasn't standardized. Allowing them to use slashes (and parsing the slashes into directories) gives them a standardized way to do organization of content which fits well with git.

Reply:
Ah gotcha. It makes sense to want to allow labs to be divided into folders, but that just means we need to change our "slug = filesystem location" mapping, not that we should change how our identifiers are formed or delimited.

Reply:
Including the semantic version with the slug and library in identifier strings is entirely new, and does require a change in how the identifier is formed.

Using a new delimiter for the version is the only way to maintain pure backwards compatibility with any existing content. While unlikely to actually exist, a slug of `12345` is currently perfectly valid. If our identifier changes to `[library]/[slug]/[version]` where library and version are optional, the string `thing/12345` becomes impossible to parse. It could mean (library: thing, slug: 12345) or (slug: thing, version: 12345) depending on the unbundler implementation, and neither is inherently more correct.

Using a new delimiter like `@` is explicit, easy to read, easy to understand, and easy to implement. Here's a regex that reliably parses the entire identifier: https://regexr.com/53eem

Reply:
I thought there already were examples of `[library]/[slug]/[version]/[revision]` somewhere but I could be making that up. It's definitely something we've talked about before, but maybe we only ended up using `libraries/[library]/content/[slug]/versions/[version]/[revisions]/revision` or whatever. I didn't think we would ever have a case where library and version were both optional, but I suppose that's possible. The other way of doing it then would be to leave `id` as just `[library]/[slug]` and add an extra `version` attribute so that we don't have to keep going with string parsing.

You're right that `@` is explicit, easy to read, easy to understand, and easy to implement - I'm not arguing against any of that. The thing to keep in mind with things like this is whether all of our systems and APIs make the *same* explicit, easy to read, easy to understand, and easy to implement decision. Since the same argument could be made about using `|` or `>` or something else we just need to be careful to make decisions like this publicly and consistently.

If this is truly the first time that we are including the semantic version with slug and library, then I guess any decision is fine, as long as we make sure that the same decision is made when this comes up in a different context.